cli/js/decode_utf8.ts

```diff
@@ -0,0 +1,134 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// `.apply` can actually take a typed array, though the type system doesn't
+// really support it, so we have to "hack" it a bit to get past some of the
+// strict type checks.
+declare global {
+  interface CallableFunction extends Function {
+    apply<T, R>(
+      this: (this: T, ...args: number[]) => R,
+      thisArg: T,
+      args: Uint16Array
+    ): R;
+  }
+}
+
+export function decodeUtf8(
+  input: Uint8Array,
+  fatal: boolean,
+  ignoreBOM: boolean
+): string {
+  let outString = "";
+
+  // Prepare a buffer so that we don't have to do a lot of string concats, which
+  // are very slow.
+  const outBufferLength: number = Math.min(1024, input.length);
+  const outBuffer = new Uint16Array(outBufferLength);
+  let outIndex = 0;
+
+  let state = 0;
+  let codepoint = 0;
+  let type: number;
+
+  let i =
+    ignoreBOM && input[0] === 0xef && input[1] === 0xbb && input[2] === 0xbf
+      ? 3
+      : 0;
+
+  for (; i < input.length; ++i) {
+    // Encoding error handling
+    if (state === 12 || (state !== 0 && (input[i] & 0xc0) !== 0x80)) {
+      if (fatal)
+        throw new TypeError(
+          `Decoder error. Invalid byte in sequence at position ${i} in data.`
+        );
+      outBuffer[outIndex++] = 0xfffd; // Replacement character
+      if (outIndex === outBufferLength) {
+        outString += String.fromCharCode.apply(null, outBuffer);
+        outIndex = 0;
+      }
+      state = 0;
+    }
+
+    // prettier-ignore
+    type = [
+       0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+       0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+       0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+       0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+       1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,  9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,
+       7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,  7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,
+       8,8,2,2,2,2,2,2,2,2,2,2,2,2,2,2,  2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
+      10,3,3,3,3,3,3,3,3,3,3,3,3,4,3,3, 11,6,6,6,5,8,8,8,8,8,8,8,8,8,8,8
+    ][input[i]];
+    codepoint =
+      state !== 0
+        ? (input[i] & 0x3f) | (codepoint << 6)
+        : (0xff >> type) & input[i];
+    // prettier-ignore
+    state = [
+       0,12,24,36,60,96,84,12,12,12,48,72, 12,12,12,12,12,12,12,12,12,12,12,12,
+      12, 0,12,12,12,12,12, 0,12, 0,12,12, 12,24,12,12,12,12,12,24,12,24,12,12,
+      12,12,12,12,12,12,12,24,12,12,12,12, 12,24,12,12,12,12,12,12,12,24,12,12,
+      12,12,12,12,12,12,12,36,12,36,12,12, 12,36,12,12,12,12,12,36,12,36,12,12,
+      12,36,12,12,12,12,12,12,12,12,12,12
+    ][state + type];
```

Comment:
Can't we gain even more perf by creating those long arrays outside of the function ?

Reply:
> Can't we gain even more perf by creating those long arrays outside of the function ?

Seems like v8 optimizes it. I wrote this decode function, and in my testing I got more performance by inlining the arrays than by having them as consts defined outside of the function.

Reply:
Yeah, V8 optimises things like this, not sure why it is more performant this way, but likely because V8 can understand 100% that there are no mutations to this arrays because they exist in a function that is 100% knowable.  In another scope, they might limit the analysis of potential mutations.  They aren't like strings and the string table, and where const means const.

Reply:
really good to know thanks !